DigitalLearningSolutions.Data/Services/PasswordResetService.cs

```diff
@@ -2,100 +2,111 @@
 {
     using System;
     using System.Collections.Generic;
-    using System.Data;
     using System.Linq;
-    using Dapper;
+    using System.Threading.Tasks;
+    using DigitalLearningSolutions.Data.DataServices;
+    using DigitalLearningSolutions.Data.Enums;
     using DigitalLearningSolutions.Data.Exceptions;
+    using DigitalLearningSolutions.Data.Models.Auth;
     using DigitalLearningSolutions.Data.Models.Email;
     using DigitalLearningSolutions.Data.Models.User;
     using Microsoft.Extensions.Logging;
     using MimeKit;
 
     public interface IPasswordResetService
     {
+        Task<bool> PasswordResetHashIsValidAsync(string emailAddress, string resetHash);
         void GenerateAndSendPasswordResetLink(string emailAddress, string baseUrl);
     }
 
     public class PasswordResetService : IPasswordResetService
     {
-        private const string AdminTableName = "AdminUsers";
-        private const string DelegatesTableName = "Candidates";
-        private const string AdminIdColumnName = "AdminID";
-        private const string DelegatesIdColumnName = "CandidateID";
-        private readonly IDbConnection connection;
+        private readonly IPasswordResetDataService passwordResetDataService;
         private readonly IEmailService emailService;
+        private readonly IClockService clockService;
         private readonly ILogger<PasswordResetService> logger;
 
         private readonly IUserService userService;
 
         public PasswordResetService(
             IUserService userService,
-            IDbConnection connection,
+            IPasswordResetDataService passwordResetDataService,
             ILogger<PasswordResetService> logger,
-            IEmailService emailService)
+            IEmailService emailService,
+            IClockService clockService)
         {
             this.userService = userService;
-            this.connection = connection;
+            this.passwordResetDataService = passwordResetDataService;
             this.logger = logger;
             this.emailService = emailService;
+            this.clockService = clockService;
+        }
+
+        public async Task<bool> PasswordResetHashIsValidAsync(string emailAddress, string resetHash)
+        {
+            var (adminUsersWithEmail, delegateUsersWithEmail) = userService.GetUsersByEmailAddress(emailAddress);
+
+            var resetPasswordIdsForEmail = adminUsersWithEmail.Select(user => user.ResetPasswordId)
+                .Concat(delegateUsersWithEmail.Select(user => user.ResetPasswordId))
+                .Where(id => id.HasValue)
+                .Select(id => id.Value);
+
+            var resetPasswordEntitiesInDb = await Task.WhenAll(
+                resetPasswordIdsForEmail.Select(id => passwordResetDataService.FindAsync(id)));
+
+            return resetPasswordEntitiesInDb.Any(
+                entity => entity != null
```

Comment:
Can this be simplified to the following (using null conditional operator)?
entity => entity?.ResetPasswordHash == resetHash
                          && clockService.UtcNow - entity?.PasswordResetDateTime < TimeSpan.FromHours(2)